* DONE find the distribution of the length of the clusters and prepare the clustering procedure
* DONE fix omp
* DONE small cases
* DONE introduce the various lambdas for taking into account of the effect of clusterization in the ratio between mirna and target absolute quantities (this is not needed!) and introduce a second lambda acting as a tuning parameter for modeling the ratio between the mirna absolute quantity and the gene absolute quantity
* DONE validate the animation with the real data
* DONE check the rate of convergence independently of the magnitude of the mirna level
* DONE plot the animation for clusters
* DONE compute the exchange for all the pairs, both the instant one and the cumulative
* DONE check the gene with id 118 (the gene indexing was wrong)
* DONE complete the code for the gaussian perturbation
* DONE find the mRNAs which are most influenced, both perturbing the base level and adding some copies to the knocked-down dataset; look both at log-ratio and at Spearman's correlation coefficient
* DONE systhematic perturbation by a fixed amount
* TODO more plots on gaussian perturbation
** TODO for the already generated data try plotting the mirna expression
** TODO try absolute perturbations
* TODO consider isoforms
* TODO stability of the method, for the expression levels, the exchange and the context (do it in the log-scale to have comparable constant variance)
* TODO find the elements which cause most of the variability in the method
* TODO check the simulation with conserved only
* TODO in cons and non-cons check if the simpler methods would work, if not how many rows do not make it possible
* TODO perturb the mRNA
* TODO adjust the amplication factor with the log
* TODO investigate the ceRNA hypothesis by perturbing few mRNAs
* TODO use TargetScan data for comparing the prediction with miRanda
* TODO include the version id in gene comparison
* REAL DATA VALIDATION
** STABILITY ANALYSIS
*** DONE make the program work with the new data format
    CLOSED: [2018-11-21 Wed 12:11]
*** DONE fix analyze_matchings_predictions.r to conform the new data format
    CLOSED: [2018-11-21 Wed 12:11]
*** DONE modify the matrix plot to support plots taken among different experiments
    CLOSED: [2018-11-21 Wed 15:28]
*** TODO make the predictions perturbing all the mirnas, to see if the ranking between different replicates coincide
*** TODO use featureCounts for finding the expression profiles of isoforms from hela data
*** TODO check the predictions when using replicates for mRNA
*** TODO look at the gene expression profiles from St. Olav server
*** TODO fix order of pertrbations in matrix plot
** TRANSFECTION
*** TODO predict the gene down-regulation for the real data and compare the predictions with pairwise targetscan scores
** VALIDATION WITH GDC LIKE IN THE ARTICLE
*** TODO look at the mirna expression profiles for HeLa cells
** FUTURE
*** TODO include the usage of version numbers
*** TODO verify that the conversion between ensembl ids and miRbase ids is correct. For instance look at ENSG00000210151
 INTERESTING BUT NOT NOW
* TODO interesting: find the correlation between the density of the entries of the interaction matrix and the quantity of exchanged mirnas and clusters
* TODO compute the weighted context score and compare it the one computed trivially
* TODO check different RPM and choose the right treshold based on different samples and of the variance
* TODO highlight the differences in the response between normal and tumor samples
NOT NOW OR NOT IMPORTANT
* TODO sites_with_scored_interactions.tsv should not contain the site type
* TODO check that "reads_per_million_miRNA_mapped" is equal to the RPM computed on "reads"
* TODO apply the changes I have been suggested in Python code
* TODO implicit first order approximation
* TODO remove cluster_debugging from the history
* TODO analyze more trivial simulations
* TODO check if I am missing information about complementary strands on miRNAs
* TODO replace all the _H defines in the headers into _HPP
* TODO check that in all the loop in which the index is an unsigned int, unsigned long or unsigned long long, I am not using an "i - 1" statement (or equivalent) in the condition
PHD APPLICATION
* TODO find the real GDC data and build the interaction graph on the difference of expressed mRNAs (consider using average levels of tumor and normal, to study the average differences)
TO FINISH THE THESIS
* TODO build the multiple linear regression model
BONUS
* TODO use the tool to find (with bruteforce and stepwise approach) to find a set of artificial sequences to insert in the cell colture to minimise the differences in down-regulation between normal and tumor cells
* TODO in generating the data use the mirbase-id instead of the miRNA family
* TODO why is the 3' UTR of ~/particular_case not available in ensembl?
